Mathlib/Dynamics/SymbolicDynamics/Basic.lean

```diff
@@ -0,0 +1,731 @@
+composed of:
+* `carrier`: the underlying set of allowed configurations.
+* `isClosed`: the set is topologically closed in `A^G`.
+* `shiftInvariant`: the set is invariant under all right-translation shifts
+  `(shift g)`. -/
+structure Subshift where
+  /-- The underlying set of configurations (additive group version). -/
+  carrier : Set (G → A)
+  /-- Closedness of `carrier`. -/
+  isClosed : IsClosed carrier
+  /-- Shift invariance of `carrier` for the additive shift `shift`. -/
+  shiftInvariant : ∀ g : G, ∀ x ∈ carrier, shift g x ∈ carrier
+
+end SubshiftDef
+
+section MulSubshiftDef
+variable (A : Type*) [TopologicalSpace A]
+variable (G : Type*) [Monoid G]
+
+/-- Multiplicative version of the definition of subshift. -/
+@[to_additive]
+structure MulSubshift where
+  /-- The underlying set of configurations. -/
+  carrier : Set (G → A)
+  /-- Closedness of `carrier`. -/
+  isClosed : IsClosed carrier
+  /-- Shift invariance of `carrier`. -/
+  shiftInvariant : ∀ g : G, ∀ x ∈ carrier, mulShift g x ∈ carrier
+
+end MulSubshiftDef
+
+
+/-- Example: the **full shift** on alphabet `A` over the multiplicative monoid `G`.
+
+It is the subshift whose underlying set is the set of all configurations
+`G → A`.
+-/
+@[to_additive SymbolicDynamics.FullShift.fullShift
+/-- Example: the **full shift** on alphabet `A` over the additive monoid `G`.
+
+It is the subshift whose underlying set is the set of all configurations
+`G → A`.
+-/]
+def mulFullShift (A G) [TopologicalSpace A] [Monoid G] : MulSubshift A G where
+  carrier := Set.univ
+  isClosed := isClosed_univ
+  shiftInvariant := by
+    intro _ _ _
+    simp
+
+/-- A *pattern* is a finite configuration in the full shift `A^G`.
+
+It consists of:
+* a finite subset `support : Finset G` of coordinates, called the support of `p`;
+* an assignment `data : support → A` of symbols to each point in the support.
+
+Intuitively, a pattern is a "partial configuration" (like a finite word),
+specifying finitely many values of a configuration in `G → A`. Patterns are
+the basic building blocks used to define subshifts via forbidden configurations.
+Note that each pattern corresponds to a cylinder, which is the set of configurations
+which agree with this pattern on its support. -/
+structure Pattern (A : Type*) (G : Type*) where
+  /-- Finite support of the pattern. -/
+  support : Finset G
+  /-- The value (symbol) at each point of the support. -/
+  data : support → A
+
+section Dominos
+
+variable (G : Type*) [Monoid G]
+
+/-- A *domino* is a pattern supported on exactly two positions (sometimes
+also called cells) `{i, j}`, specifying the value `ai` at `i` and the value `aj` at `j`.
+
+In symbolic dynamics, dominos (two-cell patterns) are the simplest nontrivial
+building blocks: they describe local adjacency conditions between two sites
+of a configuration. -/
+def Domino {A : Type*} (i j : G) (ai aj : A) : Pattern A G := by
```

Comment:
```suggestion
def domino {A : Type*} (i j : G) (ai aj : A) : Pattern A G := by
```
The rule is: if you're defining a type (including Props and structures), then upper case. Otherwise, lower case.